### ISSUE TITLE
Precision does not increase after a few epochs. Not even with more train images.

### ISSUE THREAD

**Issue by MichaelRinger** [ultralytics/ultralytics#2999]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am currently attempting to determine the minimum number of annotated images required to develop an automatic annotation model. Since many objects in the dataset are occluded, I have only annotated those with a readable product name (class). My objective is to achieve a precision and recall of over 90% using approximately 150 images. In an initial experiment with 50 training images, I achieved a precision of around 80%. However, when I increased the training set to 100 images, the precision and recall of the model did not improve. In fact, in both runs, there was no increase in precision and recall after the 6th epoch.

### Datasets:
**small**: https://app.roboflow.com/michael-ringer/cream-small/4
**large**: https://universe.roboflow.com/michael-ringer/cream-f08xf/dataset/14

### Model settings:
model=yolov8s
task=segment
epochs=50
lr0=0.003
imgsz=640
batch=4
pretrained=true
cache=ram
seed=0

### Metrics:
**small**:
![image](https://github.com/ultralytics/ultralytics/assets/33952961/daedd974-6e0c-4c56-87ca-abd6277947a8)

**large**:
![image](https://github.com/ultralytics/ultralytics/assets/33952961/e89dd332-6d99-46f8-bd88-5acadb57e68d)

Thanks for the help in advance.

**Comment by glenn-jocher** [comment]
@MichaelRinger hello! Based on the provided information, it looks like you may be experiencing an issue with the model's ability to improve precision over time. There could be many factors at play here, but one aspect to consider is the quality and diversity of the annotated images. You mentioned that many objects in the dataset are occluded so it may be helpful to include these type of objects in the annotations to help the model learn to correctly identify them. 

Additionally, the yolov8s model you are using is a smaller model, so it may not have the capacity to learn the level of detail you are looking for. You can try upgrading to a larger model such as yolov8m or yolov8l and see if that improves your precision and recall metrics.

Lastly, keep in mind that precision and recall are affected by dataset characteristics such as object frequency and image complexity, as well as model parameters such as architecture, initialization, and optimization settings. So, it may be helpful to experiment with different model parameters, loss functions and learning rates to see if any changes result in improvements to precision and recall. I hope this helps! Let us know if you have any other questions.

**Comment by MichaelRinger** [comment]
Thank you for your answer, @glenn-jocher . I appreciate your feedback. I have separately annotated objects that are visible and occluded, creating the classes "cream" and "cream-occluded."

Since this is my first real project with YOLO, I am uncertain if the annotation rules I have created make sense. Therefore, I would like to present them to you for further evaluation.

### Annotation rules:
For each class, there is an occluded subclass by default.

**Normal:**
- Products that are visually recognizable, usually by their imprinted product name
- Maximum of up to half of the object can be truncated or occluded

### Occluded:
- Minimum of up to half of the object can be cut off or occluded
- The imprinted product name must not be visible
- This class can be confused with other classes to maintain high recall. For example, a yogurt cup may look similar to a cream cup.

**Omit:**
- Objects that are very small or strongly hidden
- Maximum of 20% of the object can be recognizable
- If the object can be easily confused with other products

I was also not that strict when annotating objects that are distinguishable between other objects. For example 
Regarding your recommendations on achieving better precision, I have conducted the following experiments using the same configuration mentioned in my question thread:

### Experiment 1: Dataset large with occluded
![image](https://github.com/ultralytics/ultralytics/assets/33952961/d6338470-81e8-4415-9f2d-58505974dade)

**Result:** There is a slight improvement in precision, but the difference is not significant enough for me to distinguish it clearly between.

### Experiment 2: Dataset large YOLOv8l
![image](https://github.com/ultralytics/ultralytics/assets/33952961/b7c94558-30f9-409c-9bbd-7f9b239ce64c)

**Result:** Overall, YOLOv8l performs better.

I have also tested other hyperparameters, but I believe the issue currently lies with the data (annotations) since the precision curve is jagged and does not show significant improvement at the initial epochs.

One problem that I encountered is that by only annotating the normal ones, which require their product name (class) need to be readable, a lot of otherwise good visible examples are not annotated. For example in this image are the creams that belong to the normal class yellow and the purple creams belong to the occluded class. The right creams name is not visible but other than that it is pretty good visible. So does this annotation rule makes sense?
![image](https://github.com/ultralytics/ultralytics/assets/33952961/7e56ee49-402f-4efa-9797-6659f989dd08)

Another problem came along is that sometimes even for me objects are not easy to distinguish that need to be annotated with the occluded class. Most of the occluded ones could also be yogurt cups because the product name is not readable, but because they are in the cream shelf I annotated them as a occluded cream.
![image](https://github.com/ultralytics/ultralytics/assets/33952961/16659c2f-14e3-408e-9e52-4291094fb019)

**Comment by glenn-jocher** [comment]
Hello @MichaelRinger! Thank you for providing detailed information and your experiment results. Regarding your annotation rules, they seem to be reasonable and should be able to help the model learn to distinguish between normal and occluded objects effectively. Regarding your concern about missing some visible objects by only annotating those with a readable product name, you could consider adding them as a separate class to the dataset, with a name like "visible-cream". This would help the model learn to recognize these objects even when the product name is not visible.

Regarding your second question, it is possible to create a custom training loop with YOLOv8, which can facilitate better debugging. However, this would require some coding on your part. One approach you could take is to modify the [train.py](https://github.com/ultralytics/yolov5/blob/master/train.py) file to include the functionality for saving predictions on additional images. You could also modify the code to log the loss for each predicted image, which would allow you to identify the images with the highest loss values and filter them out.

Lastly, regarding your experiment results, it seems that increasing the dataset size and upgrading to YOLOv8l has led to some improvements in precision. Keep in mind that precision and recall are affected by many factors, and it might be necessary to experiment with different hyperparameters and loss functions to achieve optimal results.

I hope this helps! Let us know if you have any further questions or concerns.

**Comment by MichaelRinger** [comment]
Apologies for the delayed response, @glenn-jocher. I've been occupied with annotating seven additional classes, each comprising an average of 120 images. These classes primarily consist of fridge items such as butter, eggs, and cheese. I utilized the same configuration as previously mentioned, but with a reduced learning rate of 0.0001. Consequently, the precision has now dropped to 70%, while the recall stands at 60%. This represents a decrease of approximately 20% compared to when only one class with 120 images was used. I'm curious if such a substantial decline in precision and recall is typical when expanding the dataset with additional classes, as I have done.
